File: apps/api/src/gibbs/api/copilot.py
"""Sidebar copilot endpoints: chats (persisted conversations) and a streaming
reply. The reply is Server-Sent Events over POST (the browser reads it with
fetch + ReadableStream; EventSource cannot POST)."""

from __future__ import annotations

import json
import logging

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy.orm import selectinload
from sse_starlette.sse import EventSourceResponse

from ..agent.llm import model_available
from ..config import get_settings
from ..copilot import CopilotContext, CopilotDeps, build_copilot_agent, stream_reply
from ..copilot.agent import INSTRUCTIONS
from ..copilot.registry import COPILOT_AGENT_NAME, load_agent_definition
from ..copilot.transcript import (
    FORM_PATCH,
    history_from_rows,
    split_new_messages,
    transcript_from_rows,
)
from ..db.base import get_session_factory
from ..db.models import AuthUser, CopilotChat, CopilotMessage, CopilotToolCall, MessageSkill
from ..schemas import (
    CopilotChatCreate,
    CopilotChatRead,
    CopilotChatSummary,
    CopilotMessageCreate,
    CopilotStatus,
)
from .auth import require_user
from .deps import get_session

logger = logging.getLogger("gibbs.copilot")
router = APIRouter(prefix="/copilot", tags=["copilot"])


async def _messages(session: AsyncSession, chat_id: str) -> list[CopilotMessage]:
    rows = await session.execute(
        select(CopilotMessage)
        .where(CopilotMessage.chat_id == chat_id)
        .options(selectinload(CopilotMessage.tool_calls))
        .order_by(CopilotMessage.created_at, CopilotMessage.id)
    )
    return list(rows.scalars().all())


async def _read(session: AsyncSession, chat: CopilotChat) -> CopilotChatRead:
    return CopilotChatRead(
        id=str(chat.id),
        title=chat.title,
        transcript=transcript_from_rows(await _messages(session, chat.id)),
        created_at=chat.created_at,
        updated_at=chat.updated_at,
    )


async def _owned_chat(session: AsyncSession, chat_id: str, user: AuthUser) -> CopilotChat:
    chat = await session.get(CopilotChat, chat_id)
    if chat is None or chat.user_id != user.id:
        raise HTTPException(status_code=404, detail="chat not found")
    return chat


@router.get("/status", response_model=CopilotStatus)
async def copilot_status():
    """Whether the copilot can answer (its model's provider key is configured)."""
    model = get_settings().agent_model
    ok, reason = model_available(model)
    return CopilotStatus(available=ok, model=model, reason=reason)


@router.get("/chats", response_model=list[CopilotChatSummary])
async def list_chats(
    session: AsyncSession = Depends(get_session),
    user: AuthUser = Depends(require_user),
):
    """The user's chats, most recently active first."""
    rows = await session.execute(
        select(CopilotChat)
        .where(CopilotChat.user_id == user.id)
        .order_by(CopilotChat.updated_at.desc())
        .limit(50)
    )
    return [
        CopilotChatSummary(id=str(c.id), title=c.title, created_at=c.created_at, updated_at=c.updated_at)
        for c in rows.scalars().all()
    ]


@router.post("/chats", response_model=CopilotChatRead, status_code=201)
async def create_chat(
    body: CopilotChatCreate,
    session: AsyncSession = Depends(get_session),
    user: AuthUser = Depends(require_user),
):
    chat = CopilotChat(user_id=user.id, title=body.title or "New chat")
    session.add(chat)
    await session.commit()
    await session.refresh(chat)
    return await _read(session, chat)


@router.get("/chats/{chat_id}", response_model=CopilotChatRead)
async def get_chat(
    chat_id: str,
    session: AsyncSession = Depends(get_session),
    user: AuthUser = Depends(require_user),
):
    return await _read(session, await _owned_chat(session, chat_id, user))


@router.delete("/chats/{chat_id}", status_code=204)
async def delete_chat(
    chat_id: str,
    session: AsyncSession = Depends(get_session),
    user: AuthUser = Depends(require_user),
):
    chat = await _owned_chat(session, chat_id, user)
    await session.delete(chat)
    await session.commit()


@router.post("/chats/{chat_id}/messages")
async def send_message(
    chat_id: str,
    body: CopilotMessageCreate,
    session: AsyncSession = Depends(get_session),
    user: AuthUser = Depends(require_user),
):
    """Send a message and stream the copilot's reply as SSE events:
    `text` {delta}, `tool_call` {id,name,args}, `tool_result` {id,name,ok},
    `patch` {patch,rationale} (a form proposal), `done` {transcript}, `error` {detail}.

    The user row is written first; the assistant row and its tool calls are
    written when the run completes, so a failed run leaves only the question.
    """
    chat = await _owned_chat(session, chat_id, user)
    settings = get_settings()
    definition = await load_agent_definition(session, COPILOT_AGENT_NAME, INSTRUCTIONS)
    model = definition.model or settings.agent_model
    ok, reason = model_available(model)
    if not ok:
        raise HTTPException(status_code=503, detail=f"copilot unavailable: {reason}")
    context = CopilotContext(
        page=body.context.page if body.context.page in ("new_campaign", "campaign", "other") else "other",
        campaign_id=body.context.campaign_id,
        form=body.context.form,
    )
    history = history_from_rows(await _messages(session, chat.id))
    user_row = CopilotMessage(
        chat_id=chat.id,
        role="user",
        message=body.content,
        page_context=body.context.model_dump(mode="json"),
    )
    session.add(user_row)
    if chat.title == "New chat":
        chat.title = body.content.strip().splitlines()[0][:100]
    await session.commit()
    user_row_id = user_row.id
    prompt = body.content
    agent = build_copilot_agent(model, definition)

    async def generator():
        # The request-scoped session may be closed before a streaming response
        # finishes; run the reply on its own session.
        async with get_session_factory()() as run_session:
            deps = CopilotDeps(
                session=run_session, user_id=user.id, context=context, definition=definition
            )
            try:
                async for event in stream_reply(agent, prompt, history, deps):
                    if event["type"] != "done":
                        yield {"event": event["type"], "data": json.dumps(event, default=str)}
                        continue
                    text, calls = split_new_messages(event["new_messages"])
                    patch = deps.patches[-1] if deps.patches else None
                    assistant = CopilotMessage(
                        chat_id=chat_id,
                        role="assistant",
                        message=text,
                        total_tokens=event.get("total_tokens") or 0,
                        trace_id=event.get("trace_id"),
                        component_type=FORM_PATCH if patch else None,
                        component_data=patch,
                        tool_calls=[CopilotToolCall(**c) for c in calls],
                        skills=[
                            MessageSkill(skill_set_id=ref.skill_set_id, skill_name=ref.name)
                            for ref in deps.loaded_skills
                        ],
                    )
                    run_session.add(assistant)
                    stored_chat = await run_session.get(CopilotChat, chat_id)
                    stored_chat.updated_at = assistant.created_at or stored_chat.updated_at
                    await run_session.commit()
                    payload = {
                        "type": "done",
                        "transcript": transcript_from_rows(await _messages(run_session, chat_id)),
                    }
                    yield {"event": "done", "data": json.dumps(payload, default=str)}
            except Exception as exc:  # surface provider/tool errors to the sidebar
                logger.exception("copilot reply failed")
                # The question stays in the timeline without an answer; remove it so
                # the scientist can resend without a dangling turn in the history.
                row = await run_session.get(CopilotMessage, user_row_id)
                if row is not None:
                    await run_session.delete(row)
                    await run_session.commit()
                yield {"event": "error", "data": json.dumps({"type": "error", "detail": str(exc)})}

    return EventSourceResponse(generator())

File: apps/web/db/schema/copilot.ts
import { index, integer, jsonb, serial, text, timestamp, uuid } from "drizzle-orm/pg-core";
import { agent } from "./schemas";
import { user } from "./auth";

/**
 * Copilot chat model: a chat owns an ordered timeline of messages, and each
 * assistant message owns the tool calls the model made while producing it.
 * The Pydantic AI message history is rebuilt from these rows for each turn;
 * nothing is stored as an opaque history blob.
 */

const ts = (name: string) => timestamp(name, { withTimezone: true, mode: "date" });

export const chat = agent
  .table(
    "chat",
    {
      id: uuid("id").defaultRandom().primaryKey(),
      userId: text("user_id")
        .notNull()
        .references(() => user.id, { onDelete: "cascade" }),
      title: text("title").notNull(),
      createdAt: ts("created_at").notNull().defaultNow(),
      updatedAt: ts("updated_at").notNull().defaultNow(),
    },
    (t) => [index("chat_user_id_updated_at_idx").on(t.userId, t.updatedAt)],
  )
  .enableRLS();

export const messages = agent
  .table(
    "messages",
    {
      id: serial("id").primaryKey(),
      chatId: uuid("chat_id")
        .notNull()
        .references(() => chat.id, { onDelete: "cascade" }),
      /** "user" | "assistant" */
      role: text("role").notNull(),
      message: text("message").notNull(),
      totalTokens: integer("total_tokens").notNull().default(0),
      /** Display-only structured payload, e.g. "form_patch" with {patch, rationale}. */
      componentType: text("component_type"),
      componentData: jsonb("component_data").$type<Record<string, unknown>>(),
      /** OpenTelemetry / Arize trace id of the agent run that produced this message. */
      traceId: text("trace_id"),
      /** What the scientist was looking at when they sent this (user rows only). */
      pageContext: jsonb("page_context").$type<Record<string, unknown>>(),
      createdAt: ts("created_at").notNull().defaultNow(),
      updatedAt: ts("updated_at").notNull().defaultNow(),
    },
    (t) => [
      // Every chat read is "this chat's messages in timeline order".
      index("messages_chat_id_created_at_id_idx").on(t.chatId, t.createdAt, t.id),
    ],
  )
  .enableRLS();

export const toolCall = agent
  .table(
    "tool_call",
    {
      id: serial("id").primaryKey(),
      messageId: integer("message_id")
        .notNull()
        .references(() => messages.id, { onDelete: "cascade" }),
      /** The model's tool_call_id, needed to rebuild the history the model saw. */
      callId: text("call_id").notNull(),
      name: text("name").notNull(),
      /** Tool arguments as JSON text. */
      arguments: text("arguments").notNull(),
      /** Tool return value as JSON text (or the retry/error message). */
      output: text("output"),
      /** "ok" | "error" | "pending" */
      status: text("status").notNull().default("pending"),
      /** Position within the assistant turn. */
      position: integer("position").notNull().default(0),
      createdAt: ts("created_at").notNull().defaultNow(),
      updatedAt: ts("updated_at").notNull().defaultNow(),
    },
    (t) => [index("tool_call_message_id_idx").on(t.messageId)],
  )
  .enableRLS();

export type Chat = typeof chat.$inferSelect;
export type NewChat = typeof chat.$inferInsert;
export type Message = typeof messages.$inferSelect;
export type NewMessage = typeof messages.$inferInsert;
export type ToolCall = typeof toolCall.$inferSelect;
export type NewToolCall = typeof toolCall.$inferInsert;
